File: 260115/돌 잘 치우기/clear-stones-well.py
from collections import deque

n, k, m = map(int, input().split())
grid = [list(map(int, input().split())) for _ in range(n)]

starts = []
for _ in range(k):
    ri, ci = map(int, input().split())
    starts.append((ri - 1, ci - 1))

# 2. 제거 후보 선정 (모든 벽 '1'의 위치)
walls = [(i, j) for i in range(n) for j in range(n) if grid[i][j] == 1]
ans = 0
dirs = [(-1, 0), (1, 0), (0, -1), (0, 1)]

def get_count(selected_walls):
    q = deque(starts)
    visited = [[False] * n for _ in range(n)]
    for sr, sc in starts:
        visited[sr][sc] = True
    
    remove_set = set(selected_walls)
    count = 0
    
    while q:
        curr_r, curr_c = q.popleft()
        count += 1
        
        for dr, dc in dirs:
            nr, nc = curr_r + dr, curr_c + dc
            if 0 <= nr < n and 0 <= nc < n and not visited[nr][nc]:
                if grid[nr][nc] == 0 or (nr, nc) in remove_set:
                    visited[nr][nc] = True
                    q.append((nr, nc))
    return count

def backtracking(idx, selected):
    global ans
    if len(selected) == m:
        ans = max(ans, get_count(selected))
        return
    
    if idx == len(walls):
        ans = max(ans, get_count(selected))
        return

    selected.append(walls[idx])
    backtracking(idx + 1, selected)
    selected.pop()
    
    backtracking(idx + 1, selected)

actual_m = min(m, len(walls))
backtracking(0, [])

print(ans)

File: 260529/돌 잘 치우기/clear-stones-well.java
import java.util.Scanner;
import java.util.ArrayDeque;
import java.util.ArrayList;

public class Main {

    static int n, k, m;
    static int[][] grid;
    static int[][] starts;
    static int[][] walls;
    static int wallCount;
    static int ans = 0;
    static int[] drs = {1,-1,0,0};
    static int[] dcs = {0,0,1,-1};

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        n = sc.nextInt();
        k = sc.nextInt();
        m = sc.nextInt();
        grid = new int[n][n];
        for (int i = 0; i < n; i++)
            for (int j = 0; j < n; j++)
                grid[i][j] = sc.nextInt();

        starts = new int[k][2];
        for (int i = 0; i < k; i++) {
            starts[i][0] = sc.nextInt() - 1;
            starts[i][1] = sc.nextInt() - 1;
        }

        ArrayList<int[]> wallList = new ArrayList<>();
        for (int i = 0; i < n; i++)
            for (int j = 0; j < n; j++)
                if (grid[i][j] == 1)
                    wallList.add(new int[]{i, j});

        walls = wallList.toArray(new int[0][]);
        wallCount = walls.length;

        int actualM = Math.min(m, wallCount);
        backtracking(0, new boolean[wallCount], 0, actualM);

        System.out.println(ans);
    }

    static void backtracking(int idx, boolean[] selected, int selectedCount, int actualM) {
        if (selectedCount == actualM) {
            ans = Math.max(ans, getCount(selected));
            return;
        }
        if (idx == wallCount) {
            ans = Math.max(ans, getCount(selected));
            return;
        }

        selected[idx] = true;
        backtracking(idx + 1, selected, selectedCount + 1, actualM);
        selected[idx] = false;

        backtracking(idx + 1, selected, selectedCount, actualM);
    }

    static int getCount(boolean[] selected) {
        boolean[][] visited = new boolean[n][n];
        ArrayDeque<int[]> q = new ArrayDeque<>();

        for (int[] s : starts) {
            if (!visited[s[0]][s[1]]) {
                visited[s[0]][s[1]] = true;
                q.addLast(new int[]{s[0], s[1]});
            }
        }

        boolean[][] removeSet = new boolean[n][n];
        for (int i = 0; i < wallCount; i++)
            if (selected[i])
                removeSet[walls[i][0]][walls[i][1]] = true;

        int count = 0;
        while (!q.isEmpty()) {
            int[] cur = q.pollFirst();
            count++;

            for (int j = 0; j < 4; j++) {
                int nr = cur[0] + drs[j];
                int nc = cur[1] + dcs[j];

                if (0<=nr && nr<n && 0<=nc && nc<n && !visited[nr][nc]) {
                    if (grid[nr][nc] == 0 || removeSet[nr][nc]) {
                        visited[nr][nc] = true;
                        q.addLast(new int[]{nr, nc});
                    }
                }
            }
        }
        return count;
    }
}